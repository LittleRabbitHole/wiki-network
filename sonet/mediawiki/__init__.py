# -*- coding: utf-8 -*-
##########################################################################
#                                                                        #
#  This program is free software; you can redistribute it and/or modify  #
#  it under the terms of the GNU General Public License as published by  #
#  the Free Software Foundation; version 2 of the License.               #
#                                                                        #
#  This program is distributed in the hope that it will be useful,       #
#  but WITHOUT ANY WARRANTY; without even the implied warranty of        #
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         #
#  GNU General Public License for more details.                          #
#                                                                        #
##########################################################################

import re
import sys
from socket import inet_ntoa, inet_aton, error
from urllib import urlopen
from collections import namedtuple

from pageprocessor import PageProcessor, HistoryPageProcessor

try:
    import json
except ImportError:
    import simplejson as json


def fast_iter(context, func):
    """
    Use this function with etree.iterparse().

    See http://www.ibm.com/developerworks/xml/library/x-hiperfparse/ for doc.
    """
    for _, elem in context:
        func(elem)
        elem.clear()
        while elem.getprevious() is not None:
            del elem.getparent()[0]
    del context


def isip(s):
    """
    >>> isip("192.168.1.1")
    True
    >>> isip("not-an-ip")
    False
    """
    try:
        return inet_ntoa(inet_aton(s)) == s
    except error:
        return False


def isSoftRedirect(rawWikiText):
    r"""
    Find if the page starts with a soft redirect template

    >>> isSoftRedirect("{{softredirect|User:bot}}")
    True
    >>> isSoftRedirect("\n\n{{\nsoftredirect \n |  :en:User talk:bot}}")
    True
    >>> isSoftRedirect("{{ softredirect}}")
    False
    >>> isSoftRedirect("some text {{softredirect|:en:User talk:bot}}")
    False
    """
    rex = r'^[\n ]*{{[\n ]*softredirect[\n ]*\|[^}\n]*\}\}'
    return re.match(rex, rawWikiText) is not None

def is_archive(pagetitle):
    """
    Test whether a page is an archive or not
    (i.e. it contains a '/' in its title

    >>> is_archive('7_July_2005_London_bombings')
    False
    >>> is_archive('7_July_2005_London_bombings/Archive_2')
    True
    >>> is_archive('7_July_2005_London_bombings\/Archive_2')
    True
    >>> is_archive('7_July_2005_London_bombings/Archive_2/Some/thing/else')
    True
    """
    return bool(pagetitle.count('/'))

def isHardRedirect(rawWikiText):
    """
    >>> isHardRedirect("   #REDIRECT [[User:me]]")
    True
    >>> isHardRedirect("[[User:me]]")
    False
    """
    rex = r'[\n ]*#REDIRECT[\n ]*\[\[[^]]*\]\]'
    return re.match(rex, rawWikiText) is not None

## re_cache is a mutable, so it keeps state through function calls
def getCollaborators(rawWikiText, search, lang=None, re_cache = {}):
    """
    Search for regular expression containing [[User:username|anchor text]] and
    count a new message from username to the owner of the page. It also works
    on localized versions of the wikipedia, for example in the Italian
    Wikipedia it searches for
    [[Utente:username|anchor text]]

    We choose to get [[User:username|anchor text]] and not the
    User_discussion:username link for the following reason: signatures can be
    personalized.

    We rely on policies for signatures in the different Wikipedias.
    In English Wikipedia, see
    http://en.wikipedia.org/wiki/Wikipedia:Signatures#Links "Signatures must
    include at least one internal link to your user page, user talk page, or
    contributions page; this allows other editors easy access to your talk
    page and contributions log. The lack of such a link is widely viewed as
    obstructive." In Italian wikipedia, see
    http://it.wikipedia.org/wiki/Aiuto:Personalizzare_la_firma#Personalizzare_la_firma
    "La firma deve sempre contenere un wikilink alla pagina utente e/o alla
    propria pagina di discussione."

    >>> getCollaborators( \
            'd [[User:you|mee:-)e/e]] d [[User:me]][[utente:me]]', \
                         ('Utente', 'User'))
    {u'Me': 2, u'You': 1}
    >>> getCollaborators('[[User:you', ('Utente', 'User'))
    {}
    >>> getCollaborators('[[Utente:me/archive|archive]]', ('Utente', 'User'))
    {}
    >>> getCollaborators('[[:vec:Utente:me|or you]]', ('Utente', 'User'), \
            'vec')
    {u'Me': 1}

    """
    if lang:
        search += tuple([":%s:%s" % (lang, s) for s in search])
    rex = r'\[\[(%s):([^/]*?)[|\]][^\]]*\]' % ('|'.join(search),)
    try:
        matches = re_cache[rex].finditer(rawWikiText)
    except KeyError:
        re_cache[rex] = re.compile(rex, re.IGNORECASE)
        matches = re_cache[rex].finditer(rawWikiText)

    weights = dict()
    for u in matches:
        ##TODO: fare un test per controllare che questo continui a funzionare
        ##      (nomi utenti strani da correggere con capfirst e replace)
        u2 = u.group(2)
        if not u2:
            print >>sys.stderr, 'getCollaborators: empty username found'
            continue
        un = capfirst(unicode(u2).replace('_', ' '))
        weights[un] = weights.get(un, 0) + 1

    return weights


##TODO: add doctests
def getTemplates(rawWikiText):
    rex = '\{\{(\{?[^\}\|\{]*)'
    matches = re.finditer(rex, rawWikiText)

    weights = dict()
    for tm in matches:
        t = tm.group(1)
        weights[t] = weights.get(t, 0) + 1

    return weights


#def getWords(rawWikiText):
#    import nltk

def addGroupAttribute(g, lang, group='bot'):
    url = ('http://%s.wikipedia.org/w/api.php?action=query&list=allusers'+
           '&augroup=%s&aulimit=500&format=json') % (lang, group)

    start = None
    while True:
        if start:
            url += '&aufrom=%s' % (start,)
        furl = urlopen(url)
        res = json.load(furl)

        if not res.has_key('query') or not res['query']['allusers']:
            print 'Group %s has errors or has no users' % group
            g.vs[group] = [None,]*len(g.vs)
            return

        for user in res['query']['allusers']:
            print user['name'].encode('utf-8')
            try:
                g.vs.select(username=user['name'].encode('utf-8')
                            )[0][group] = True
            except IndexError:
                pass

        if res.has_key('query-continue'):
            start = res['query-continue']['allusers']['aufrom']
        else:
            break

    return


def addBlockedAttribute(g, lang):
    g.vs['blocked'] = [None,]*len(g.vs)
    url = base_url = ('http://%s.wikipedia.org/w/api.php?action=query&list='+
                      'blocks&bklimit=500&format=json') % ( lang, )

    start = None
    while True:
        if start:
            url = '%s&bkstart=%s' % (base_url, start)
        print "BLOCKED USERS: url = %s" % url
        furl = urlopen(url)
        res = json.load(furl)

        if not res.has_key('query') or not res['query']['blocks']:
            print 'No blocked users'
            return

        bk_list = []
        for block in res['query']['blocks']:
            if not block.has_key('user'):
                continue
            print block['user'].encode('utf-8')
            try:
                bk_list.append(block['user'].encode('utf-8'))
            except IndexError:
                pass

        bk_vs = g.vs.select(username_in=bk_list)
        bk_vs['blocked'] = (True,)*len(bk_vs)

        if res.has_key('query-continue'):
            start = res['query-continue']['blocks']['bkstart']
        else:
            break

    return


def getTags(src, tags='page,title,revision,text'):
    # find namespace (eg: http://www.mediawiki.org/xml/export-0.3/)
    try:
        root = src.readline()
        ns = unicode(re.findall(r'xmlns="([^"]*)', root)[0])

        tag_prefix = u'{%s}' % ns

        tag = {}
        for t in tags.split(','):
            tag[t] = tag_prefix + unicode(t)
    finally:
        src.seek(0)

    return tag


def getTranslations(src):
    try:
        counter = 0
        translation = {}

        while 1:
            line = src.readline()
            if not line: break
            keys = re.findall(
                r'<namespace key="(\d+)"[^>]*>([^<]*)</namespace>',
                line)
            for key, ns in keys:
                if key == '1':
                    translation['Talk'] = unicode(ns, 'utf-8')
                if key == '2':
                    translation['User'] = unicode(ns, 'utf-8')
                elif key == '3':
                    translation['User talk'] = unicode(ns, 'utf-8')
                elif key == '4':
                    translation['Wikipedia'] = unicode(ns, 'utf-8')

            counter += 1
            if counter > 50:
                break
    finally:
        src.seek(0)

    return translation

def getNamespaces(src):
    try:
        counter = 0
        namespaces = []

        while 1:
            line = src.readline()
            if not line: break
            keys = re.findall(r'<namespace key="(\d+)"[^>]*>([^<]*)</namespace>',
                              line)
            for key, ns in keys:
                namespaces.append((key, ns))

            counter += 1
            if counter > 40:
                break
    finally:
        src.seek(0)

    return namespaces


def explode_dump_filename(fn):
    """
    >>> explode_dump_filename( \
            "/tmp/itwiki-20100218-pages-meta-current.xml.bz2")
    ('it', '20100218', '-pages-meta-current')
    """
    from os.path import split

    s = split(fn)[1] #filename with extension
    res = re.search('(.*?)wiki[\-]*-(\d{8})([^.]*)', s)
    return (res.group(1), res.group(2), res.group(3))


def capfirst(s):
    """
    Given a string, it returns the same string with the first letter capitlized

    >>> capfirst("test")
    'Test'
    """
    return s[0].upper() + s[1:]


def count_renames(lang):
    url = base_url = ('http://%s.wikipedia.org/w/api.php?action=query&list='+\
                      'logevents&letype=renameuser&lelimit=500&leprop='+ \
                      'title|type|user|timestamp|comment|details&format=json'
                      ) % ( lang, )
    counter = 0
    start = None
    while True:
        if start:
            url = '%s&lestart=%s' % (base_url, start)
        furl = urlopen(url)
        res = json.load(furl)

        if not res.has_key('query') or not res['query']['logevents']:
            print 'No logs'
            return

        counter += len(res['query']['logevents'])

        if res.has_key('query-continue'):
            start = res['query-continue']['logevents']['lestart']
        else:
            break
        print counter

    return counter

Message = namedtuple('Message', 'time welcome')