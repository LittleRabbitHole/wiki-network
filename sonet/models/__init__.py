#!/usr/bin/env python

##########################################################################
#                                                                        #
#  This program is free software; you can redistribute it and/or modify  #
#  it under the terms of the GNU General Public License as published by  #
#  the Free Software Foundation; version 2 of the License.               #
#                                                                        #
#  This program is distributed in the hope that it will be useful,       #
#  but WITHOUT ANY WARRANTY; without even the implied warranty of        #
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         #
#  GNU General Public License for more details.                          #
#                                                                        #
##########################################################################

from sqlalchemy import Table, MetaData, create_engine, Column, Integer, \
     String, Boolean, Sequence

def get_engine():
    return create_engine(
        'postgresql+psycopg2://pgtharpe:tharpetharpe@tharpe/research_wiki'
    )

def get_events_table(engine=None, metadata=None):
    if engine is None:
        engine = get_engine()
    if metadata is None:
        metadata = MetaData()
        metadata.bind = engine

    events = Table('wikinetwork_wikievent', metadata,
                   Column('id', Integer,
                          Sequence('wikinetwork_wikievent_id_seq'),
                          primary_key=True),
                   Column('title', String),
                   Column('lang', String),
                   Column('desired', Boolean),
                   Column('data', String),
                   Column('talk', Boolean)
                   )
    conn = engine.connect()

    return (events, conn)
