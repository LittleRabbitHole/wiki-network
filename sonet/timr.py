import sys
from time import time

class Timr(object):
    counters = {}
    main_counter = None
    name = None
    
    def __init__(self, name=None):
        self.name = name
        
    def __enter__(self):
        self.start(self.name)
    
    def __exit__(self, exc_type, exc_value, traceback):
        self.stop(self.name)

    def start(self, name=None):
        if name:
            print >> sys.stderr, "START %s" % name
            self.counters[name] = time()
        else:
            print >> sys.stderr, "START TIMER"
            self.main_counter = time()

    def stop(self, name=None):
        if name:
            print >> sys.stderr, "STOP %s: %6f" % (name,  time() - self.counters[name])
        else:
            print >> sys.stderr, "STOP TIMER: %6f" % (time() - self.main_counter, )
