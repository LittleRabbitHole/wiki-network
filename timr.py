import sys
from time import time

class Timr(object):
    counters = {}
    main_counter = None

    def start(self, name=None):
        if name:
            print >> sys.stderr, "START %s" % name
            self.counters[name] = time()
        else:
            print >> sys.stderr, "START TIMER"
            self.main_counter = time()

    def stop(self, name=None):
        if name:
            print >> sys.stderr, "STOP %s: %6f" % (name,  time() - self.counters[name])
        else:
            print >> sys.stderr, "STOP TIMER: %6f" % (time() - self.counters[name], )
