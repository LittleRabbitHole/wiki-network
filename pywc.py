#!/usr/bin/env python

########################################################################
#                                                                      #
# Copyleft Federico "fox" Scrinzi (fox91 at anche dot no)              #
#                                                                      #
# pyWC is free software; you can redistribute it and/or modify         #
# it under the terms of the GNU General Public License as published by #
# the Free Software Foundation; version 2 of the License.              #
#                                                                      #
# pyWC program is distributed in the hope that it will be useful,      #
# but WITHOUT ANY WARRANTY; without even the implied warranty of       #
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the         #
# GNU General Public License for more details.                         #
#                                                                      #
########################################################################

import csv
csv.field_size_limit(100000000)
import sys
import logging
try:
    import re2 as re
except ImportError:
    logging.warn("pyre2 not available. It's gonna be a long job")
    import re
from sonet.timr import Timr


class PyWC:
    """
    PyWC is a python class for word counting and text analisys.
    """
    # Global proprieties (of the whole source file)
    categories = None        # Dictionary's categories
    keywords = None          # Dictionary's keywords/regex
    delimiter = "\t"         # CSV delimiter
    quotechar = '"'          # CSV quotechar
    csv_out = sys.stdout     # CSV output
    queue = []               # Flushing queue
    max_char_limit = 100000  # Max chars per line
    ignorecols = []          # List of columns of the src file to ignore
    csv_writer = None        # Csv writer handler
    id_col = 0               # Number of the id column
    dic_regex = False        # Use dictionary made of regex
    flush_n = 100            # Number of pieces of text to store
    clean_wiki = True        # Clean wiki syntax
    clean_html = True        # Clean HTML

    clean_regex = (
        (re.compile(r"(^|\s)(\w\.)+"), ""),
        (re.compile(r"[\:|;|8|=|-]+[a-z](\s|$)", re.IGNORECASE), ""))

    # TODO TODO TODO TODO TODO
    clean_wiki_regex = (
        (re.compile(r"(?:https?://)(?:[\w]+\.)(?:\.?[\w]{2,})+(\S+)?"), ""),
        (re.compile(r"\[{1,2}([^\:\|]+?)\]{1,2}", re.DOTALL), r"\1"),
        (re.compile(r"\[{1,2}.+?[\||\:]([^\|^\:]+?)\]{1,2}",
                    re.DOTALL), r"\1"),
        (re.compile(r"\[{1,2}.+?\]{1,2}", re.DOTALL), ""),
        (re.compile(r"\{{1,3}.+?\}{1,3}", re.DOTALL), ""),
        #(re.compile(r"[\w|\s]+:\w+(.+?\])?", re.U), ""))
        (re.compile(r"\|(.+)?(\s+?)?=(\s+?)?(.+)?"), ""))

    # Stripping HTML tags and comments
    clean_html_regex = ((re.compile(r"<\!--.+?-->", re.DOTALL), ""),
                        (re.compile(r"<.+?>"), ""),
                        (re.compile(r"\&\w+;"), ""))

    cond_exp_regex = (re.compile(r"<([\w']+)>(\w+)(\/(\w+)?)?"),
                      re.compile(r"\(([\w\s]+)\)(\w+)(\/(\w+)?)?"))

    # Local proprieties (of every column of the source file)
    _id = None         # Line ID
    _results = None    # Dictionary where keys are cat ids and
                       # values are counters
    _total = None      # List of numbers of total words per column
    _text = None       # Current text to analize
    _next_word = None  # Next word that has to be analized
    _prev_cat = None   # Categories of the last word that has been analized
                       # (useful for conditional exps)
    _counter = 0       # Generic counter of how many pieces of
                       # text have been analized
    _keys = None

    def __init__(self, **kwargs):
        self.__dict__ = kwargs

    def delattrs(self, attrs):
        """
        Frees memory deleting useless attributes of the object
        """
        for attr in attrs:
            try:
                delattr(self, attr)
            except AttributeError:
                pass

    def _gen_keyword(self, content):
        """
        Generator for self.keywords (dictionary made of of regexps
        as keys and thier categories as values)
        """
        for line in content[2].split("\n")[1:-1]:
            # Comments start with //
            if line and not line.startswith("//"):
                line = line.split("\t")
                # If not using a dictionary made of regexps
                # it fixes the keyword for regexping
                # "^" is added at the beginning of every keyword
                # If keyword doesn't ends with "*", a "$" is added
                # bad -> ^bad$ matches "bad" but not "badass"
                # bad* -> ^bad matches "bad" and "badass"
                if not self.dic_regex:
                    line[0] = "".join(["^", line[0]])
                    try:
                        if (line[0][-1] == "*"):
                            line[0] = line[0][:-1]
                        else:
                            line[0] = "".join([line[0], "$"])
                    except IndexError:
                        continue
                yield (re.compile(line[0]), line[1:])

    def set_dic(self, dic):
        """
        Receives as input the dictionary filename.
        Reads the dictionary file and populates self.categories and
        self.keywords
        """
        f = open(dic, 'r')
        content = f.read()
        content = content.split("%")
        if len(content) != 3:
            raise ValueError("Invalid dic file")

        # Creates a dictionary where category ids are the keys
        # and category names are the values.
        # Splits content at first by new line, then by tab
        self.categories = dict((line.split("\t") \
                for line in content[1].split("\n")[1:-1] if line))

        # Creates a dictionary where the compiled regex is the key
        # and category ids are the values
        self.keywords = dict(x for x in self._gen_keyword(content))

    def flush(self):
        """
        Writes everything which is in the queue in the csv output file
        """
        self.csv_writer.writerows(self.queue)
        self.queue = []

    def save(self):
        """
        Saves current piece of text that has been analized to the queue
        """
        tmp = {"id": self._id,
               "total": self._total,
               "text": self._text}
        # Join of self.categories and self._results values
        for k, v in ((self.categories[x], self._results[x]) \
                     for x in self.categories):
            tmp[k] = v
        self.queue.append(tmp)
        del tmp
        self._counter += 1
        if self._counter % self.flush_n == 0:
            logging.info("### Flushing: %d", self._counter)
            self.flush()

    def parse_word(self, word):
        """
        Parses a single word with the dictionary of regexps
        (self.keywords). For every regex that matches, it
        increments every category they belong to in self._result
        """
        cat = []
        for regex in self.keywords:
            if regex.search(word):
                for i in self.keywords[regex]:
                    res = self.cond_exp_regex[0].match(i)
                    if res:
                        if self._next_word == res.group(1):
                            cat.append(res.group(2))
                        elif res.group(4):
                            cat.append(res.group(4))
                        continue

                    res = self.cond_exp_regex[1].match(i)
                    if res:
                        if True in [c in self._prev_cat \
                                    for c in res.group(1).split(" ")]:
                            cat.append(res.group(2))
                        elif res.group(4):
                            cat.append(res.group(4))
                        continue

                    # If dictionary contains trailing tabs,
                    # '' keys are saved. It skips them.
                    if i:
                        cat.append(i)

        for c in cat:
            try:
                self._results[c] += 1
            except KeyError:
                logging.warn("Invalid category id %s", c)
        self._total += 1
        self._prev_cat = cat

    def clean_wiki_syntax(self):
        """
        Cleans self._text from wiki syntax
        """
        # TODO FILTER WIKI SYNTAX
        for regex, replace in self.clean_wiki_regex:
            self._text = regex.sub(replace, self._text)

    def clean_html_syntax(self):
        """
        Cleans self._text from HTML tags and comments
        """
        for regex, replace in self.clean_html_regex:
            self._text = regex.sub(replace, self._text)

    def clean_text(self):
        """
        Cleans self._text from emoticons and acronyms
        """
        for regex, replace in self.clean_regex:
            self._text = regex.sub(replace, self._text)

    def parse_col(self, col):
        """
        Reads a single cell of the csv file. It splits it
        into words and gives them to self.parse_word
        """
        self.delattrs(("_results", "_total", "_text",
                       "_prev_word", "_prev cat"))
        self._text = col
        logging.info("--------PRIMA-----------")
        logging.info(self._text)
        logging.info("-------------------")
        self.clean_text()
        if self.clean_wiki:
            self.clean_wiki_syntax()
        if self.clean_html:
            self.clean_html_syntax()
        logging.info("--------DOPO------------")
        logging.info(self._text)
        logging.info("-------------------")
        self._results = {}
        for k in self.categories:
            self._results[k] = 0
        self._total = 0
        rwords = re.compile("[\w']+")
        # create a list of words (_no_ numbers)
        words = [word for word in rwords.findall(col) if not word.isdigit()]
        for i, word in enumerate(words):
            try:
                self._next_word = words[i+1]
            except IndexError:
                self._next_word = ""
            self.parse_word(word)
        self.save()

    def parse_line(self, line):
        """
        Reads a single line of the csv file.
        Sets self._id and gives the cells that are not in the ignore
        list to self.parse_col
        """
        self.delattrs(("_id"))
        self._id = line[self.id_col]
        for i, col in enumerate(line):
            if len(col) <= self.max_char_limit:
                if i != self.id_col and not i in self.ignorecols:
                    self.parse_col(col)
            else:
                logging.warn(" Line %d:%d skipped "
                             "because longer than %d chars",
                             self._counter, i, self.max_char_limit)

    def start(self, src):
        """
        It starts the file processing.
        To obtain a sensible output is recommended to run self.set_dic()
        before.
        It writes the output csv header and reads every line, passing
        it to self.parse_line
        """

        # Creates a list of category names sorted by their ID.
        # Useful because Python dictionaries are not sorted objects!
        # Sorting like TAWC
        try:
            cat_names = [x[1] for x in sorted([(int(a), b) for a, b in \
                                               self.categories.items()])]
        except ValueError:
            cat_names = [x[1] for x in sorted(self.categories.items())]

        self._keys = ["id"] + cat_names + ["total", "text"]
        self.csv_writer = csv.DictWriter(self.csv_out,
                                         delimiter=self.delimiter,
                                         fieldnames=self._keys,
                                         quotechar=self.quotechar)
        self.csv_writer.writeheader()
        csv_reader = csv.reader(src, delimiter=self.delimiter)
        for line in csv_reader:
            self.parse_line(line)


def main():
    import optparse
    p = optparse.OptionParser(
        usage="usage: %prog [options] dic_file input_file")
    p.add_option('-v', action="store_true", dest="verbose", default=False,
                 help="Verbose output (like timings)")
    p.add_option('-c', '--charlimit', action="store", dest="charlimit",
                 type="int", default=100000,
                 help="Maximim characters per line (default=100000)")
    p.add_option('-i', '--ignorecols', action="store", dest="ignorecols",
                 help="Coulmns numbers of the source file to ignore" + \
                      "(comma separated and starting from 0)")
    p.add_option('-I', '--id', action="store", dest="id_col", type="int",
                 help="Id column number (starting from 0)", default=0)
    p.add_option('-r', action="store_true", dest="regex", default=False,
                 help="Use a dictionary composed by regex (default=false)")
    p.add_option('-f', "--flush", action="store", dest="flush", type="int",
                 default=100,
                 help="Flushing to output every N pieces of text")
    p.add_option('-o', "--output", action="store", dest="output",
                 help="Output file (default=STDOUT)")
    opts, files = p.parse_args()
    if len(files) != 2:
        p.error("Wrong parameters")

    if opts.verbose:
        logging.basicConfig(stream=sys.stderr,
                            level=logging.DEBUG,
                            format='%(asctime)s %(levelname)s %(message)s',
                            datefmt='%Y-%m-%d %H:%M:%S')

    t = PyWC()
    t.max_char_limit = opts.charlimit
    if opts.ignorecols:
        t.ignorecols = [int(x) for x in opts.ignorecols.split(",")]
    t.id_col = opts.id_col
    t.dic_regex = opts.regex
    t.flush_n = opts.flush
    if opts.output is not None:
        t.csv_out = open(opts.output, 'w')

    t.set_dic(files[0])
    src = open(files[1], 'r')
    with Timr("Processing"):
        t.start(src)

    t.flush()


if __name__ == "__main__":
    main()
