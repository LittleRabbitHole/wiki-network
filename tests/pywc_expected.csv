id	po	super	ne	other	total	text
a123	1	1	0	0	8	Hello there, today is a really good day
a123	2	0	1	0	6	Today is a so great day
4444	0	0	2	0	6	Today is a bad baddd day
4444	1	0	0	0	4	Today I feel good
aaaa	1	0	0	0	5	it ain't a good day
1111	2	1	0	0	5	I definitely like that game
1234	0	0	2	0	4	I don't like him
1234	0	0	0	1	7	what kind of software do you use?
1111	1	0	0	0	5	I met a kind girl
1234	1	1	0	0	5	That girl is really kind
